Classes/EventListener/EventListenerInvoker.php

```diff
@@ -18,81 +18,84 @@
+        $eventStream = $this->eventStore->load($streamName);
         foreach ($eventStream as $eventEnvelope) {
             try {
                 $this->applyEvent($listener, $eventEnvelope);
+                $highestAppliedSequenceNumber = $eventEnvelope->getRawEvent()->getSequenceNumber();
             } catch (EventCouldNotBeAppliedException $exception) {
+                $appliedEventsStorage->saveHighestAppliedSequenceNumber($highestAppliedSequenceNumber);
                 $appliedEventsStorage->releaseHighestAppliedSequenceNumber();
                 throw $exception;
             }
-            $appliedEventsStorage->saveHighestAppliedSequenceNumber($eventEnvelope->getRawEvent()->getSequenceNumber());
             if ($progressCallback !== null) {
                 $progressCallback($eventEnvelope);
             }
         }
+        $appliedEventsStorage->saveHighestAppliedSequenceNumber($highestAppliedSequenceNumber);
         $appliedEventsStorage->releaseHighestAppliedSequenceNumber();
     }
 
     /**
      * @param EventListenerInterface $listener
-     * @return EventStore
+     * @param \Closure $progressCallback
+     * @throws EventCouldNotBeAppliedException
      */
-    public function getEventStoreForEventListener(EventListenerInterface $listener): EventStore
+    public function catchUp(EventListenerInterface $listener, \Closure $progressCallback = null): void
     {
-        $listenerClassName = \get_class($listener);
-        $eventStoreIdentifier = $this->eventListenersConfiguration[$listenerClassName]['eventStore'] ?? 'default';
-        try {
-            return $this->eventStoreFactory->create($eventStoreIdentifier);
-        } catch (\InvalidArgumentException $exception) {
-            throw new \RuntimeException(sprintf('Failed to build Event Store for listener "%s": %s', $listenerClassName, $exception->getMessage()), 1570191582, $exception);
+        $appliedEventsStorage = $this->getAppliedEventsStorageForListener($listener);
+        $highestAppliedSequenceNumber = $appliedEventsStorage->reserveHighestAppliedEventSequenceNumber();
+        $streamName = $listener instanceof StreamAwareEventListenerInterface ? $listener::listensToStream() : StreamName::all();
+        $eventStream = $this->eventStore->load($streamName, $highestAppliedSequenceNumber + 1);
+        foreach ($eventStream as $eventEnvelope) {
+            try {
+                $this->applyEvent($listener, $eventEnvelope);
+            } catch (EventCouldNotBeAppliedException $exception) {
+                $appliedEventsStorage->releaseHighestAppliedSequenceNumber();
+                throw $exception;
+            }
+            $appliedEventsStorage->saveHighestAppliedSequenceNumber($eventEnvelope->getRawEvent()->getSequenceNumber());
+            if ($progressCallback !== null) {
+                $progressCallback($eventEnvelope);
+            }
         }
+        $appliedEventsStorage->releaseHighestAppliedSequenceNumber();
```

Comment:
TODO: discuss transaction stuff here

Reply:
OK, Shoot :)
Is it related to this change?

Reply:
nope :)

Reply:
OK then, let's create a separate issue (or wait until this poses an issue?)